addons/hr_timesheet/models/hr_timesheet.py

```diff
@@ -183,24 +221,22 @@ def _timesheet_preprocess(self, vals):
             Overrride this to compute on the fly some field that can not be computed fields.
             :param values: dict values for `create`or `write`.
         """
-        # project implies analytic account
-        if vals.get('project_id') and not vals.get('account_id'):
-            project = self.env['project.project'].browse(vals.get('project_id'))
+        project = self.env['project.project'].browse(vals.get('project_id', False))
+        task = self.env['project.task'].browse(vals.get('task_id', False))
+        # task implies project
+        if task and not project:
+            project = task.project_id
+            if not project:
+                raise ValidationError(_('You cannot create a timesheet on a private task.'))
```

Comment:
I don't think the private task exists in 14.0. (to check) 🤔

Reply:
@xavierbol  Indeed, at least no `ir.rule` is made like in 16.0 
https://github.com/odoo/odoo/blob/ce60532b0c5c77a80038298a27550ab5f9eeb012/addons/project/security/project_security.xml#L150-L151
Do we still raise an error `You cannot create a timesheet on a task not linked to a project.` ?

Reply:
If we are sure a task will be linked to a project then we can assume this error will never arrive. 

For me, it is not really a problem to let that error. It is just a bit useless if all tasks are linked to a project in that version. 🙂

Reply:
@xavierbol In fact, from the frontend it is ok because we make the project_id field required in XML, but not in the backend :
https://github.com/odoo/odoo/blob/20ec8f6a4684573d0190424cc03a969797bc4b47/addons/project/views/project_views.xml#L659
https://github.com/odoo/odoo/blob/a51058d17ad70764f795e272d6578b33842dba1c/addons/project/models/project.py#L597-L599
However, the field `account_id` is required in backend, so also in frontend.
https://github.com/odoo/odoo/blob/4214f05d454bca2b60fda3a288d529c098e84f77/addons/analytic/models/analytic_account.py#L187
Therefore, if we try to create an AAL without giving an `account_id` or the `project_id` (where we check and get the `account_id`), we have a SQL error:

odoo-shell:
```py
task_copy = self.env['project.task'].browse(1).copy()
task_copy.project_id = False

aal = self.env['account.analytic.line'].create({
    'name': 'test',
    'task_id': task_copy.id,
})
```
=> error  : `psycopg2.errors.NotNullViolation: null value in column "account_id" violates not-null constraint`

Should I adapt the code so that we raise an error if no `account_id` is set instead (it will however allow creating an AAL without project...) or should I let the error on the task ( so that the `account_id` will be managed on the `project_id`)? 
https://github.com/odoo/odoo/blob/22c8e3810de0a09b4c566b92d670b344a267f261/addons/hr_timesheet/models/hr_timesheet.py#L232-L236

Reply:
Nope, I think when no project is linked to the AAL then it is not a timesheet and our code should not handle the AAL. We already manage enough things. 😄